File: frontend/src/api/jobs.ts
import { ResearchJob, JobStatus } from '../types';
import api from './axios';

function normalizeJobStatus(status: string): JobStatus {
  return status.toLowerCase() as JobStatus;
}

function normalizeJob(job: ResearchJob): ResearchJob {
  return { ...job, status: normalizeJobStatus(job.status) };
}

export async function createJob(question: string): Promise<ResearchJob> {
  const { data } = await api.post<ResearchJob>('/jobs/', { question });
  return normalizeJob(data);
}

export async function listJobs(): Promise<ResearchJob[]> {
  const { data } = await api.get<ResearchJob[]>('/jobs/');
  return data.map(normalizeJob);
}

export async function getJob(id: string): Promise<ResearchJob> {
  const { data } = await api.get<ResearchJob>(`/jobs/${id}`);
  return normalizeJob(data);
}

export async function deleteJob(id: string): Promise<void> {
  await api.delete(`/jobs/${id}`);
}

File: backend/app/routers/jobs.py
from datetime import datetime, timezone
from typing import List
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from ..core.deps import get_current_user
from ..database import get_db
from ..models.job import JobStatus, ResearchJob
from ..models.user import User
from ..schemas.job import JobCreate, JobResponse
from ..tasks.research import run_research_job

router = APIRouter(prefix="/jobs", tags=["jobs"])


@router.post("/", response_model=JobResponse, status_code=201)
def create_job(
    data: JobCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    job = ResearchJob(
        user_id=current_user.id,
        question=data.question.strip(),
        status=JobStatus.PENDING,
    )
    db.add(job)
    db.commit()
    db.refresh(job)

    # Dispatch to Celery
    task = run_research_job.delay(str(job.id), data.question.strip())

    job.celery_task_id = task.id
    db.commit()
    db.refresh(job)

    return job


@router.get("/", response_model=List[JobResponse])
def list_jobs(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return (
        db.query(ResearchJob)
        .filter(ResearchJob.user_id == current_user.id)
        .order_by(ResearchJob.created_at.desc())
        .all()
    )


@router.get("/{job_id}", response_model=JobResponse)
def get_job(
    job_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    job = (
        db.query(ResearchJob)
        .filter(ResearchJob.id == job_id, ResearchJob.user_id == current_user.id)
        .first()
    )
    if not job:
        raise HTTPException(status_code=404, detail="Job not found")
    return job
